clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -335,6 +335,15 @@ public void poll(long timeout) throws IOException {
         // we use the time at the end of select to ensure that we don't close any connections that
         // have just been processed in pollSelectionKeys
         maybeCloseOldestConnection(endSelect);
+
+        Iterator<String> connectedIt = connected.iterator();
```

Comment:
@hachikuji @ijuma Agreed to avoid inconsistency in the first place. And it may be a little better to do it in close rather than in doClose, for doClose maybe a delayed operation and close is the first place to notify the ConnectionStates.